top-level: add goPlatform for Go packages

tie opened the pull request:
## Description of changes

This change adds `go` attribute/argument for the platforms returned by `lib.systems.elaborate`, similar to `rustc`. This attribute is used for top-level `goPlatform` attributes that further elaborates platforms into GOOS/GOARCH and other Go-specific variables, potentially taking Go version being used into account. Go build infrastructure is updated to use `goPlatform` instead of using GOOS/GOARCH/etc exposed from the `go` package.

See also https://github.com/NixOS/nixpkgs/pull/256505#issuecomment-1731935611

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

tie committed:
lib.systems: elaborate Go platforms

This change adds `go` attribute and argument for the
lib.systems.elaborate function, and therefore to the
localSystem/crossSystem arguments of Nixpkgs.

This is similar to `rustc` attribute and allows fine-tuning Go builds,
though this would be implemented in subsequent commits. For example:

  import <nixpkgs> {
    crossSystem = (import <nixpkgs/lib>).systems.examples.gnu32 // {
      go.GO386 = "softfloat";
    };
  }

tie committed:
build-support/go: add makeGoPlatform and Go platform concept

This change adds top-level goPlatform attribute and updates
buildGo{Module,Package} to use goPlatform instead of go package
attributes for GOOS/GOARCH and other variables.

tie committed:
go: use goPlatform.{build,host,target}.env

tie commented:
Result of `nixpkgs-review pr 256761` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

tie committed:
doc/go: add section about cross-compilation

tie committed:
rl-2311: add note about goPlatfrom

tie commented:
FWIW goPlatform name may be too verbose (`hostPlatform.goPlatform`), I’m more in favor of `hostPlatform.go`, but I’d like to delay the rename until I address review comments (if any) because *grep*ping for goPlatform is easier 😅

Ericson2314 commented:
Calling it `go` sounds good. In general, I'd recommend doing what @alyssais did for Rust.

tie commented:
@Ericson2314, do you mean `rustPlatform` infrastructure?

tie commented:
- Renamed `goPlatform` attribute/argument to `go` for lib.systems.elaborate/localSystem/crossSystem.
- Added `goPlatform` top-level attribute to avoid implementing platform-to-Go-env mapping logic in `lib.systems`. This also potentially allows adding checks and other special cases based on the Go version being used.

zowoq commented on pkgs/build-support/go/module.nix:
```diff
@@ -149,10 +153,16 @@ let
-    inherit (go) GOOS GOARCH;
+    env = (args.env or { }) // goPlatform.env.host // {
+      inherit CGO_ENABLED GO111MODULE GOTOOLCHAIN;
+      GOFLAGS = lib.concatStringsSep " " (
```
Why is this being changed to use `concatStringsSep` and `lib.optional`?

tie replied:
`concatStringsSep` because `env` does not accept lists.
https://github.com/NixOS/nixpkgs/blob/74970765262f0958d124b5b43afdd45e8285d527/pkgs/stdenv/generic/make-derivation.nix#L490-L491

`lib.optional` — no particular reason, I’ll change it back to `lib.optionals`, I think I’ve changed that accidentally in one place and then though that it was `lib.optional` all along 🥲

zowoq commented:
`ofborg` is adding the `clean-up` label because of this:
```
5 packages removed:
kind (†0.20.0) promtail (†2.9.1) slsa-verifier (†2.4.0) zabbix-agent2 (†5.0.37) zabbix-agent2 (†6.0.21)
```

Note that `ofborg` will not remove the `clean-up` label automatically, need to remove it manually and make sure `ofborg` doesn't add it again.

tie commented:
>5 packages removed

This is my bad, `nixpkgs-review` missed these for some reason. These packages set environment variables (not via `env`) that are overridden via `env`, so `mkDerivation` complains. To be fair, these variables were likely ignored before.

tie committed:
promtail: do not set CGO_ENABLED explicitly

tie committed:
slsa-verifier,kind: do not set -trimpath explicitly

We pass -trimpath when allowGoReference is false (default).

tie committed:
zabbix.agent2: do not set GOOS/GOARCH explicitly

This is no-op, GOOS and GOARCH are set from buildGoModule.

tie committed:
slsa-verifier: do not set GO111MODULE explicitly

GO111MODULE is set to "on" by default in buildGoModule.

tie commented:
Pushed minor changes that should improve UX:
- For `buildGo{Module,Package}`, `env` passed in arguments has higher precedence than what we set.
- Swapped `goPlatform.env.{build,host,target}` to `goPlatform.{build,host,target}.env`.

Ericson2314 commented:
We got rid of `rustPlatform`; we don't want a `goPlatform`. I'll try to work with @alyssais to explain how we got to the current design for Rust so mistakes don't have to be repeated/relearned.

tie commented:
Marking this as a draft for now until I have time work on this again (hopefully in a few days/weeks). I think I have a better understanding of the design decisions behind Nixpkgs now.

In particular,
- Go provide setup hooks (and standalone fetchGoModules) to allow composing Go with other languages/tools (e.g. similar to Node.js infrastructure).
- buildGo{Module,Package} should be a convenience wrapper for common Go packaging patterns.
